fixes #7757: in oci schema fix query results row keys case

nineinchnick opened the pull request:
fixes #7757: in oci schema fix query results row keys case when PDO::ATTR_CASE is set to PDO::CASE_LOWER

nineinchnick committed:
fixes #7757: in oci schema fix query results row keys case when PDO::ATTR_CASE is set to PDO::CASE_LOWER
added test to check support for setting PDO::ATTR_CASE and fetching table schemas
fixed fetching table schemas for mysql when PDO::ATTR_CASE is set
added tests for oci
fixed fetching composite fks for oci
improvements in oci schema parsing
removed autoIncrement detection fro oci and added test to verify that
implement batchInsert for oci
fix detecting IntegrityException for oci
fixed creating raw sql by skipping object and resource params
fix command test failing for sqlite

samdark commented:
What if case is neutral?

nineinchnick commented:
Then the results contains keys as specified in the query, so uppercase.

I admin I have not run this. I'll try to install Oracle at home because I'm getting more interested in it.

samdark commented:
Great.

nineinchnick commented:
I forgot to add - `PDO::CASE_NATURAL` is the default value for `PDO::ATTR_CASE`.

nineinchnick commented:
Ohh and it's a good idea to add a unit test for this. It could also check other databases.

samdark commented:
Agree.

nineinchnick commented:
Added `oci` tests, but the whole `oci` schema implementation is a total mess. Column types are not detected properly. I'm going to fix that before adding case tests.

nineinchnick commented:
I feel like the tests I wrote here are some schema fixes should be moved to another PR. Can anybody take a look?

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -276,6 +293,9 @@ protected function findTableNames($schema = '')
         $rows = $command->queryAll();
         $names = [];
         foreach ($rows as $row) {
+            if ($this->db->slavePdo->getAttribute(\PDO::ATTR_CASE) == \PDO::CASE_LOWER) {
```
`===`?

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -146,12 +137,16 @@ protected function findColumns($table)
         }
 
         foreach ($columns as $column) {
+            if ($this->db->slavePdo->getAttribute(\PDO::ATTR_CASE) == \PDO::CASE_LOWER) {
```
`===`?

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -219,7 +215,15 @@ protected function createColumn($column)
+                $defaultValue = $column['DATA_DEFAULT'];
+                if ($defaultValue !== null) {
+                    if (($len=strlen($defaultValue)) > 2 && $defaultValue[0] === "'" && $defaultValue[$len-1] === "'") {
+                        $defaultValue = substr($column['DATA_DEFAULT'], 1, -1);
```
Could it be multibyte?

nineinchnick replied:
We only check if the value is surrounded with single quotes, which can't be multibyte. I don't think it matters if the rest of the value is.

nineinchnick replied:
I was also thinking about trying to detect numerics here and if the default value is neither a string nor numeric it could be wrapper as `db\Expression`.

When processing datatypes this `Schema` removes default values for timestamps. I don't know why, should this be fixed?

samdark replied:
There's `$defaultValue[$len-1]` which could go wrong for multibyte strings because of `strlen($defaultValue)`.

nineinchnick replied:
Doesn't `strlen` just count bytes instead of chars? If it does, the last byte would always be a quote.

nineinchnick replied:
I just didn't want to assume the value is in utf.

nineinchnick replied:
The task to perform here is: check if value is surrounded by single quotes regardless of the value's encoding.

samdark replied:
Automatic wrapping with `db\Expression` should be done carefully since it could easily result in security issues.

nineinchnick replied:
Why? It's the default value expression from the db schema, not user input. Examples:

* numerics: `12.34`
* strings: `'something'`
* expressions: `to_timestamp('value','format')` and `CURRENT_TIMESTAMP`.

samdark replied:
Right.

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -246,12 +250,25 @@ protected function findConstraints($table)
         $command = $this->db->createCommand($sql);
+        $constraints = [];
         foreach ($command->queryAll() as $row) {
+            if ($this->db->slavePdo->getAttribute(\PDO::ATTR_CASE) == \PDO::CASE_LOWER) {
```
`===`?

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -219,7 +215,15 @@ protected function createColumn($column)
-                $c->defaultValue = $c->phpTypecast($column['DATA_DEFAULT']);
+                $defaultValue = $column['DATA_DEFAULT'];
+                if ($defaultValue !== null) {
+                    if (($len=strlen($defaultValue)) > 2 && $defaultValue[0] === "'" && $defaultValue[$len-1] === "'") {
```
Could it be multibyte?

samdark commented on tests/unit/framework/db/DatabaseTestCase.php:
```diff
@@ -19,6 +19,9 @@ protected function setUp()
         $databases = self::getParam('databases');
         $this->database = $databases[$this->driverName];
         $pdo_database = 'pdo_'.$this->driverName;
+        if ($this->driverName == 'oci') {
```
`===`?

samdark commented on tests/unit/framework/db/DatabaseTestCase.php:
```diff
@@ -71,7 +74,13 @@ public function prepareDatabase($config, $fixture, $open = true)
         $db->open();
         if ($fixture !== null) {
-            $lines = explode(';', file_get_contents($fixture));
+            if ($this->driverName == 'oci') {
```
`===`?

samdark commented:
Another PR could be useful but this one isn't too big to be reviewed.

nineinchnick commented:
I think I'm going to quote all column names in `oci.sql` to allow reusing more tests.

nineinchnick commented:
I've added more tests and implemented some features like `batchInsert` along the way.

`OracleActiveRecordTest` still doesn't pass, this is the last one.

If anybody wants to run this, here's [my pdo_oci8 driver](https://github.com/nineinchnick/pdo_oci8).

nineinchnick commented:
I'm done here

nineinchnick commented:
Thinking about it, AR tests should include loading metadata when `PDO:ATTR_CASE` is set to `PDO::CASE_UPPER`.

samdark commented:
There's at least one test failing now: https://travis-ci.org/yiisoft/yii2/jobs/56241236#L245

nineinchnick commented:
@samdark I fixed one test condition. I think there was parenthesis missing, can you take a look? Now Travis fails with Cubrid setup, is it normal?

samdark commented:
Cubrid fail is about travis setup. It is, of course, very annoying because of fail to see real errors.

samdark commented on framework/db/oci/QueryBuilder.php:
```diff
@@ -160,4 +160,61 @@ public function addForeignKey($name, $table, $columns, $refTable, $refColumns, $
+
+        $values = [];
+        foreach ($rows as $row) {
+            $vs = [];
```
Consider renaming it to something more meaningful.

nineinchnick replied:
This is copied from the parent class, not my own code. If this to be refactored, it should be changed in other schemas, like sqlite.

samdark replied:
OK. Let's leave it as is for now.

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -146,12 +145,16 @@ protected function findColumns($table)
                 $table->sequenceName = $this->getTableSequenceName($table->name);
-                $c->autoIncrement = true;
+                // autoIncrement cannot be reliably detected, ex. with composite PKs
+                //$c->autoIncrement = $table->sequenceName !== null;
```
Should it stay there commented?

samdark commented on framework/db/oci/Schema.php:
```diff
@@ -219,7 +224,21 @@ protected function createColumn($column)
+                        if (($len = strlen($defaultValue)) > 2 && $defaultValue[0] === "'"
+                            && $defaultValue[$len - 1] === "'"
+                        ) {
+                            $defaultValue = substr($column['DATA_DEFAULT'], 1, -1);
```
Could it be unicode?

nineinchnick replied:
We've already discussed this. Also, other schemas just trim quotes here which is invalid and could remove too much. Should I report it in another issue?

samdark replied:
Probably yes.

nineinchnick commented:
Ok I've added a test that actually tests what this issue was about and detected and fixed invalid behavior for mysql.

I can't test mssql or cubrid.

cebe commented:
> Cubrid fail is about travis setup. It is, of course, very annoying because of fail to see real errors.

This is an issue with travis, not sure what is happening as there is no error output, working for me locally. I have tried to make it more robust and not kill the whole test suite on failures:
https://github.com/yiisoft/yii2/commit/fd99fe1efabb22c0c7e019ace03152f0958a5411

If you rebase on master you should be able to see real test results.

nineinchnick commented:
@cebe I did and tests still fail. Anyhow, I'm done here with work except fixes if requested, moving on to gii issues.

cebe commented:
Yeah but now you see it is only cubrid failing, before it was the whole test suite. Your tests seem fine. The cubrid failure is unrelated and cause by an error in the setup which I have not figured out yet.

samdark commented:
@nineinchnick please add changelog and I'll merge PR.

Commit:
updated CHANGELOG.md

nineinchnick commented:
@samdark done.

samdark commented:
Thanks!

cebe commented:
@nineinchnick great job!
